server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java

```diff
@@ -126,7 +128,17 @@ private Map<String, UploadedSegmentMetadata> readLatestMetadataFile() throws IOE
 
     private Map<String, UploadedSegmentMetadata> readMetadataFile(String metadataFilename) throws IOException {
         try (IndexInput indexInput = remoteMetadataDirectory.openInput(metadataFilename, IOContext.DEFAULT)) {
-            Map<String, String> segmentMetadata = indexInput.readMapOfStrings();
+            // BufferedChecksumIndexInput reader keeps computing checksum for all the bytes we read from indexInput
+            // This checksum is then compared with checksum in footer
+            ChecksumIndexInput in = new BufferedChecksumIndexInput(indexInput);
+            CodecUtil.checkHeader(
+                in,
+                UploadedSegmentMetadata.METADATA_CODEC,
+                UploadedSegmentMetadata.CURRENT_VERSION,
+                UploadedSegmentMetadata.CURRENT_VERSION
+            );
+            Map<String, String> segmentMetadata = in.readMapOfStrings();
+            CodecUtil.checkFooter(in);
```

Comment:
Does `checkFooter` and `checkHeader` throw exception on mismatch? Can we add a test for the same?

Reply:
yes, they can throw exceptions. Let me add tests for those. There some others tests pending as well for the metadata file write flow.

Reply:
Add all required UTs